### ISSUE TITLE
successResultOf and failureResultOf should accept coroutines

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#9006]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)| @wsanchez reported|
|-|-|
|Trac ID|trac#9006|
|Type|enhancement|
|Created|2017-01-21 00:34:44Z|
|Branch|https://github.com/twisted/twisted/tree/9006-coro-result-assertions|

`twisted.trial.unittest.SynchronousTestCase.successResultOf` takes a `deferred` argument, which should be a `Deferred` object.

If given a coroutine, that should also work:

```
class SomeTests(unittest.SynchronousTestCase):
    async def someAsyncFunction():
        return 5

    def test_someAsyncFunctionReturnsFive(self):
        result = self.successResultOf(someAsyncFunction())
        self.assertEqual(result, 5)
```

Same for `twisted.trial.unittest.SynchronousTestCase.failureResultOf`.

<details><summary>Searchable metadata</summary>

```
trac-id__9006 9006
type__enhancement enhancement
reporter__wsanchez wsanchez
priority__normal normal
milestone__None None
branch__9006_coro_result_assertions 9006-coro-result-assertions
branch_author__ 
status__closed closed
resolution__fixed fixed
component__trial trial
keywords__None None
time__1484958884324682 1484958884324682
changetime__1563821720389492 1563821720389492
version__None None
owner__wsanchez wsanchez
cc__dpnova
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

Note http://twistedmatrix.com/trac/ticket/8900 - so these may not be the APIs you're looking for.

Also, why overload some existing API instead of introduce a new API?

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_3"></a>|@wsanchez commented|
|-|-|

J-P, agree that the other API is interesting here, but I'd say also here.

The reason for overloading is that `Deferred` mimics coroutines such that one can await on a deferred, same as one would a coroutine, so there are cases where the two may be mixed and matched as "async-aware thingos" and the caller isn't necessarily aware of which they have.

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_4"></a>|@wsanchez commented|
|-|-|

I find the argument about avoiding polymorphism compelling in the new API being added to `twisted.internet.defer`, as those methods are meant to tell you about the state of a deferred.

I think it's more appropriate, though, in trial for `successResultOf` to accept a generic "async thingo" and get it's current result by making the thingo into a deferred if necessary and then using that new API to check.

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_6"></a>|@wsanchez commented|
|-|-|

Also: `assertNoResult`

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_7"></a>|@wsanchez commented|
|-|-|

PR: https://github.com/twisted/twisted/pull/691

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_9"></a>|@glyph commented|
|-|-|

One of the potential problems with polymorphism here is that users of `asyncio`-based libraries may quite reasonably expect that the flavor of the coroutine returned here to be a `Future`-yielding coroutine; implicitly selecting Twisted-flavored (`Deferred`-yielding) might result in some confusion.

For the time being I'd be inclined to say that if you want to call `ensureDeferred`, it's there if you need it.

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_10"></a>|@wsanchez commented|
|-|-|

Glyph: the flavor returned by what?  The change here is to allow `successResultOf` to accept any `Awaitable` and not just `Deferred`s, so it's not polymorphic so much as changing the input type to something less restrictive.

It doesn't return a `Deferred` so I'm not sure what confusing you are referring to.

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_11"></a>|@wsanchez commented|
|-|-|

Using typing syntax, what this ticket does is change:

```
def successResultOf(d: Deferred) -> Any: ...
```

to

```
def successResultOf(d: Awaitable) -> Any: ...
```

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_12"></a>|@glyph set owner to @wsanchez|
|-|-|

Replying to [Wilfredo Sánchez Vega](#note_11):
> Using typing syntax, what this ticket does is change (... Deferred... ) to (... typing.Awaitable ...):

This syntax is illuminating.  Here's the problem, expressed more precisely: `Awaitable` is not the type you want.  _Formally_, `Awaitable` just means "implements `__await__` so it can be awaited upon", and both `Deferred` and `coroutine` satisfy that requirement.  However, _practically_ what "awaitable" idiomatically means is "implements `__await__` ''to yield `asyncio.Future`s''.

There's a generic version of `Awaitable` (`Awaitable[T]`) but what that appears to mean is "awaitable that, when awaited, evaluates to `T`.  This _really_ ought to be expressed as `Awaitable[Future[T]]` but that ship has sailed.

Of course, the only kind of `Awaitable` that we can deal with is the kind that awaits `Deferred` objects, which we have no way to express.  So what this change _actually_ gives us is the multiple-dispatch signature of:

```
def successResultOf(d: ThingPeopleUsuallyMeanByAwaitable[T]) -> useless garbage, fail to do work, traceback later probably: ...
def successResultOf(d: WeirdDeferredThingWeMeanByAwaitable[T]) -> T: ...
```

This seems like a bad situation, where people are constantly going to be trying to feed `Future`-yielding coroutines which they got from `asyncio` libraries into this function and being confused when it doesn't work, since their simple tests (`Deferred`- or nothing-awaiting coros) all appear to work fine.  I'd rather have it keep failing on our own flavor of coroutine until we have good documentation and runtime introspection that can present useful errors in a reliable way.

In short, when we get an object to `successResultOf` - how do we know whether to call `Deferred.fromFuture(ensure_future(it))` or `ensureDeferred(it)` on it?

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_13"></a>|@wsanchez commented|
|-|-|

Glyph, is your answer gonna be different if and when Python 3.7 supports the interleaving of `Deferred`s and `Future`s as I heard at PyCon is the plan?

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_14"></a>|@glyph commented|
|-|-|

Replying to [Wilfredo Sánchez Vega](#note_13):
> Glyph, is your answer gonna be different if and when Python 3.7 supports the interleaving of `Deferred`s and `Future`s as I heard at PyCon is the plan?

I confess my feelings are drifting in that direction.  Perhaps we can resolve the type distinctions by simply having a `Union` in there for now?

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_15"></a>|@wsanchez commented|
|-|-|

Glyph, I'm a little confused about that, since we don't actually use any type hints in Twisted yet, right?  (I'd like to, but working on Klein first.)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_16"></a>|@glyph commented|
|-|-|

Oh, I'm not saying we need to block on that, just that since it was a thing that had come up I was just saying I think it can be addressed.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_17"></a>|@glyph commented|
|-|-|

The issue here is that we have a thing, which is one of two states:

- If it's a Deferred, it should have already been fired
- If it's a thing that yields Deferreds when we await upon it, all those Deferreds should already have been fired.
- If it's a thing that yields Futures when we await upon it, then… somehow… those Futures need to get an `asyncio.AbstractEventLoop` passed to them which will call `call_soon` on them immediately, and when we await upon all those futures they will pass themselves to that loop.

This is what still makes it tricky.  We could make our own asyncio loop that just implements call_soon, we could create a new `Task` from that loop, but then we need to somehow give that loop to any application code which might manufacture a `Future`.  Or, we could set up a temporary event loop policy which would look up the running loop at that point.

I think this is much more tractable than I did originally, but it is still going to take some work to wire together the proof of concept.

One thing that we could do in the meanwhile is just turn the crank on the generator and then emit an extremely verbose error message if a Future is yielded. We could always add Future-yielding coroutine support in a future iteration.

This is not _quite_ the same thing as the interoperability of `Future` and `Deferred` in mutually awaitable coroutines, because that integration presupposes the existence of a "real" asyncio event loop that `Deferred` can just blindly schedule tasks with, but the possibility of that integration is what makes me think this might be doable.

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_18"></a>|@wsanchez commented|
|-|-|

OK, I totally agree with raising an exception until we figure out how to deal with a Future.

As you noted, it would be hard for someone not familiar with all of this nonsense (myself included) to make sense of the otherwise random-seeming errors that would otherwise ensue.

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)|@wsanchez set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="dpnova's avatar" src="https://avatars.githubusercontent.com/u/229943?s=50" width="50" height="50">](https://github.com/dpnova)<a name="note_21"></a>|@dpnova commented|
|-|-|

Some of the discussion here is fascinating, especially the Awaitable[Future[T]]. 

I'm currently struggling with having to write tests twisted style while the majority of the tested code reads like asyncio code. 

One particular quirk is I've not found a way to test a function that awaits a Deferred.fromFuture. ensureDeferred says its not a coroutine when it is. I feel like it's mostly related to this issue too.. Or should I open another ticket for that?

**Comment by twisted-trac** [comment]
|[<img alt="dpnova's avatar" src="https://avatars.githubusercontent.com/u/229943?s=50" width="50" height="50">](https://github.com/dpnova)<a name="note_22"></a>|@dpnova commented|
|-|-|

After a bit more playing I think [#9273](https://github.com/twisted/twisted/issues/9273) is one of the things making my life difficult.

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)|@wsanchez removed owner<br>@wsanchez set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_24"></a>|@glyph commented|
|-|-|

I think that my earlier objections are largely withdrawn, in favor of simply making a `Deferred` "task" (i.e. a coroutine run by `ensureDeferred`) be able to await upon `Future` objects and vice versa.  I outlined a possible approach to this in https://twistedmatrix.com/pipermail/twisted-python/2019-April/032280.html

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_25"></a>|@hawkowl set owner to @wsanchez|
|-|-|

I think this looks fine. The situation where you give asyncio futures, I think, is fixed by glyph's proposed changes (which from the looks are basically making ensureDeferred be smarter by returning a Deferred that can await on a Future if need be?), and don't need to be fixed here.

Apart from the lint on the changelog being wrong, merge away.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_26"></a>|@glyph commented|
|-|-|

Replying to [hawkowl](#note_25):
> I think this looks fine. The situation where you give asyncio futures, I think, is fixed by glyph's proposed changes (which from the looks are basically making ensureDeferred be smarter by returning a Deferred that can await on a Future if need be?), and don't need to be fixed here.

Agreed!

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)|@wsanchez set status to `closed`|
|-|-|
